lucene/core/src/java/org/apache/lucene/index/ExitableDirectoryReader.java

```diff
@@ -100,13 +109,157 @@ public CacheHelper getCoreCacheHelper() {
+            + queryTimeout.toString()
+            + ", PointValues=" + in
+        );
+      } else if (Thread.interrupted()) {
+        throw new ExitingReaderException("Interrupted while iterating over point values. PointValues=" + in);
+      }
+    }
+
+    @Override
+    public void intersect(IntersectVisitor visitor) throws IOException {
+      checkAndThrow();
+      in.intersect(new ExitableIntersectVisitor(visitor, queryTimeout));
+    }
+
+    @Override
+    public long estimatePointCount(IntersectVisitor visitor) {
+      checkAndThrow();
+      return in.estimatePointCount(visitor);
+    }
+
+    @Override
+    public byte[] getMinPackedValue() throws IOException {
+      checkAndThrow();
+      return in.getMinPackedValue();
+    }
+
+    @Override
+    public byte[] getMaxPackedValue() throws IOException {
+      checkAndThrow();
+      return in.getMaxPackedValue();
+    }
+
+    @Override
+    public int getNumDataDimensions() throws IOException {
+      checkAndThrow();
+      return in.getNumDataDimensions();
+    }
+
+    @Override
+    public int getNumIndexDimensions() throws IOException {
+      checkAndThrow();
+      return in.getNumIndexDimensions();
+    }
+
+    @Override
+    public int getBytesPerDimension() throws IOException {
+      checkAndThrow();
+      return in.getBytesPerDimension();
+    }
+
+    @Override
+    public long size() {
+      checkAndThrow();
+      return in.size();
+    }
+
+    @Override
+    public int getDocCount() {
+      checkAndThrow();
+      return in.getDocCount();
+    }
+  }
+
+  public static class ExitableIntersectVisitor implements PointValues.IntersectVisitor {
+
+    public static final int DEFAULT_MAX_CALLS_BEFORE_QUERY_TIMEOUT_CHECK = 10;
+
+    private final PointValues.IntersectVisitor in;
+    private final QueryTimeout queryTimeout;
+    private final int maxCallsBeforeQueryTimeoutCheck;
+    private int calls = 0;
+
+    public ExitableIntersectVisitor(PointValues.IntersectVisitor in, QueryTimeout queryTimeout) {
+      this(in, queryTimeout, DEFAULT_MAX_CALLS_BEFORE_QUERY_TIMEOUT_CHECK);
+    }
+
+    public ExitableIntersectVisitor(PointValues.IntersectVisitor in,
+                                    QueryTimeout queryTimeout, int maxCallsBeforeQueryTimeoutCheck) {
+      this.in = in;
+      this.queryTimeout = queryTimeout;
+      this.maxCallsBeforeQueryTimeoutCheck = maxCallsBeforeQueryTimeoutCheck;
+    }
+
+    /**
+     * Throws {@link ExitingReaderException} if {@link QueryTimeout#shouldExit()} returns true,
+     * or if {@link Thread#interrupted()} returns true.
+     */
+    private void checkAndThrow() {
+      if (calls++ % maxCallsBeforeQueryTimeoutCheck == 0 && queryTimeout.shouldExit()) {
+        throw new ExitingReaderException("The request took too long to intersect point values. Timeout: "
+            + queryTimeout.toString()
+            + ", PointValues=" + in
+        );
+      } else if (Thread.interrupted()) {
```

Comment:
should this also not be called everytime and moved under the `if (calls++ % maxCallsBeforeQueryTimeoutCheck == 0)` test?

Reply:
also sampling makes sense for the two visit() methods which are called in tight loops, but less for relate and grow, should we only sample visit() calls?

Reply:
My bad, outer condition fixed in 9e11f76b2ac7b9cda9f1c78ad0c1f3a5be09f435 and only sample `visit` methods fixed in 3d09426a81277a5344435f43a1942570c1a85f37.

Reply:
I was thinking of keeping the checkAndThrow in other methods too, just without sampling. Though I guess most time is going to be spent in visit() anyway so the current version isn't bad.

Reply:
My bad, fixed in 7eef7e9a8213714bdc3a90aa7fa5c5db010d5151.